adding MIR and backend support for boolean operators

eafurst opened the pull request:
Boolean operators (`and`, `or`, `xor`, and `not`) were implemented in the frontend IR, but dropped when lowered to the midend IR. This adds support for the mentioned boolean operators to the midend IR and backend code generation.

AjayBrahmakshatriya commented:
Hi @eafurst,
Thanks for the changes. We have been wanting to implement boolean operators as a feature since some time. Overall the changes look good. Few comments I have currently which I think would be better fixed before merging into the master branch - 

1. There are some lines in the diff that have some white space changes. For example - codegen_cpp.cpp (mostly the python bindings section). Can you undo those changes, so that is easy to track who last modified those lines in case something breaks? A simple checkout on the file and adding only the lines that appear at line no. 945 should be fine. 

2. Can you create some test cases for the newly implemented features? These tests can be added in the tests/python/test.py file. This lets us make sure the features you have added do not break when we make some other changes. 

Thanks again for the changes. Please let me know if you have any concerns.

yunmingzhang17 commented:
Thanks for the bug fix! The request looks good, but a few test cases would make it more future proof as we integrate more changes. 

What we usually do is 
(Step 1) add a simple graphit file in the test/input directory (https://github.com/GraphIt-DSL/graphit/tree/master/test/input) , in this example for the few boolean operators. A good example would be (https://github.com/GraphIt-DSL/graphit/blob/master/test/input/simple_if_elif_else.gt). 
(Step 2) add a test case in (https://github.com/GraphIt-DSL/graphit/blob/master/test/python/test.py), Similar to (https://github.com/GraphIt-DSL/graphit/blob/master/test/python/test.py#L296).
You can also print the output to verify the output is correct using "expect_output_val" (https://github.com/GraphIt-DSL/graphit/blob/master/test/python/test.py#L115). 

Again, let us know if you have any issues adding test cases?

eafurst commented:
Thanks for the feedback!

I've fixed the whitespace issues (I recently switched to using Atom as an editor and think that might have been causing issues).

I also added a simple test, `simple_boolean_op.gt` as requested. It is passing the compilation test that I added in `test.py` and is generating the expected C++ code and printing the expected output when I compile and run by hand. Let me know what you think - I can also add an `expect_output_val` check as well if it helps.

AjayBrahmakshatriya commented:
Hi @eafurst,
Thanks for the quick response and addressing all the comments. The changes look good. 
I will merge the pull-request.

yunmingzhang17 commented:
Thanks for the quick fix! 
If you are looking at IDEs for development, I actually use CLION (https://www.jetbrains.com/clion/) for GraphIt development. You can build the entire project and run all the tests in CLION with great debugger support as well.
